port addcmul to OpInfo

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55547 move topk to use OpInfo
* #55535 move logaddexp{2} to opinfo
* #55518 port addcdiv to opinfo
* **#55517 port addcmul to OpInfo**

Differential Revision: [D27649413](https://our.internmc.facebook.com/intern/diff/D27649413)

albanD committed:
port addcmul to OpInfo

[ghstack-poisoned]

albanD committed:
Update on "port addcmul to OpInfo"




[ghstack-poisoned]

mruberry commented on test/test_torch.py:
```diff
@@ -7080,8 +7080,7 @@ def tmp(dtype, device):
         _float_types, _cpu_types, True),
-    ('addcmul', '', _small_3d, lambda t, d: [_small_3d(t, d), _small_3d(t, d)], 1e-2, 1e-1, 1e-3,
-        torch.testing.get_all_dtypes(include_complex=True, include_bool=False)),
+    # TODO: finish the deprecation cycle for this overload of addcmul and remove this test
```
Nice attention to detail here.

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2475,6 +2500,15 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           assert_autodiffed=True,
+           skips=(
+               SkipInfo('TestCommon', 'test_variant_consistency_eager'),
+               SkipInfo('TestGradients', 'test_inplace_grad'),
```
I think the TestGradients skips can be replaced with setting supports_inplace_autograd to False:

https://github.com/pytorch/pytorch/blob/0dff0d1537a3746e80dd3a3b655ec071fe3c2466/torch/testing/_internal/common_methods_invocations.py#L166

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -2475,6 +2500,15 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+           dtypesIfCUDA=all_types_and_complex_and(torch.float16, torch.bfloat16),
+           assert_autodiffed=True,
+           skips=(
+               SkipInfo('TestCommon', 'test_variant_consistency_eager'),
```
Would you just add a comment above this skip:

```
# TODO: update sample inputs with for_inplace_variant kwarg to support this test
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -602,6 +602,31 @@ def sample_inputs_addmm(op_info, device, dtype, requires_grad, **kwargs):
         return (input, )
 
+def sample_inputs_addcmul(op_info, device, dtype, requires_grad, **kwargs):
+    test_cases = [((S, S), (S, S), (S, S)),
```
```
test_cases = (
  ((S, S), (S, S), (S, S)),
  ((S, S), (S, 1), (1, S)),
  ((1,), (S, S, 1), (1, S)),
  ((), (), ()),
  ((S, S), (), ()),
  ((), (S, S, 1), (1, S)),
)
```

Your approach is correct, but because you're (cleverly) applying the value kwarg programmatically below this doesn't need the repeated shape cases like the method_tests() entries did.

albanD replied:
Ho, I'm really bad at de-duplication... haha

mruberry reviewed (approved):
Thanks for participating in the OpInfo porting session of the Hackathon, @albanD!

This is correct but I made a few inline suggestions for refinements. Please take a look. Approving this now because they're straightforward to implement.

albanD committed:
Update on "port addcmul to OpInfo"




[ghstack-poisoned]

albanD committed:
Update on "port addcmul to OpInfo"


Differential Revision: [D27649413](https://our.internmc.facebook.com/intern/diff/D27649413)

[ghstack-poisoned]

albanD committed:
Update on "port addcmul to OpInfo"


Differential Revision: [D27649413](https://our.internmc.facebook.com/intern/diff/D27649413)

[ghstack-poisoned]

albanD committed:
Update on "port addcmul to OpInfo"


Differential Revision: [D27649413](https://our.internmc.facebook.com/intern/diff/D27649413)

[ghstack-poisoned]
